### PR TITLE
Remove the default output from Mock

### PR DESCRIPTION
**Pull request by Mcat12**
Most of the time if a mock is called with unexpected input, it is a bug and the test should fail. This PR removes the default output from `Mock` and will now fail the test if it is called with an unexpected input. Note that you can use `Matcher::Any` to still allow unknown inputs.

The current error message in the case of an unknown input is static and does not include the input or mock details. It is unknown if this can be improved without moving it to a procedural macro or using specialization on `I: Debug`.

This PR also adds a `will_return_default` method to ease writing tests now that the mock does not automatically have a default.

The removal of the default is a breaking change.

Technical details:
- Replaced `Output` with a new struct, `Given`. Mocks now store `Rule<I, O>` instead of `Rule<I, Output<O>>`.
- Added `Given::will_return_default` for `O: Default`

### CODE REVIEW

File: src/given.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    use std::sync::{Arc, Mutex};
+
+    /// When `Given::will_return` is called with an output, the corresponding
+    /// rule is added to the rules list
+    #[test]
+    fn add_rule_to_list() {
+        let rules = Arc::new(Mutex::new(Vec::new()));
+        let given = Given::new("hello", rules.clone());
+
+        assert_eq!(rules.lock().unwrap().len(), 0);
```
[the remark this answers is not in the document]
Comment (Mcat12) [reply src/given.rs]: Why should we keep the number of asserts minimum? Without this assert, we could miss a bug where `Given` inserts the rule before we call the function.

File: src/mock.rs [hunk 2]
```diff
@@ -90,8 +80,10 @@ mod test {
         }
     }
 
+    /// Given some rules, the mock should return the output in the rule when
```
[the remark this answers is not in the document]
Comment (Mcat12) [reply src/mock.rs]: Comments are a good thing. Long, hard to read method names are a bad thing. It is much easier to express what a test should do when you can write freely in a comment.

What is not clean about rust's test output?

**Comment by Mcat12** [comment]
> I think there is some unanswered questions concerning the use of default / any

What questions?

File: src/given.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nathanielsimard) [reply src/given.rs]: A test should only have one reason to fail that is not covered by another test in the same suite. That way, if something goes wrong in the implementation, not all the tests become red, only the relevant ones.

If this is important to verify that we do not add a rule initially, I would make another test that describes it like: rules are initially empty.

File: src/mock.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nathanielsimard) [reply src/mock.rs]: Yes, I also prefer to write test descriptions without constraint like snake_case, maximum method name, etc. But I also like when we can see **what** went wrong only by reading the test output.

JS have a way to write test with comments that appear in the test output BDD style. Go also have some way to do it. I was just wondering if there is a standard way of describing a test that will appear in the `cargo test` command when something goes wrong.

If nothing like that exist, I think comments are better in the case of Rust,

**Comment by nathanielsimard** [comment]
Ho sorry, my questions did not published with my review :confused:.

I think having a default value with `will_return_default`  have the same purpose as a normal `will_return` with `Any` for all inputs. I think it's better to aims for one way of doing think to avoid confusion. But maybe I miss something here...

Also, the procedural macro do not support `Matcher` at the moment. I'm unsure if we should add it by default, have another macro, or another solution to support it ?

File: src/mock.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mcat12) [reply src/mock.rs]: That style of test name writing is not supported by Rust's native test framework.

At least in my editor, I can click on any test and go immediately to the test definition, so this is no issue for me. Even in cases where this is not the case, the test name is usually enough to know what it was testing, at least generally.

**Comment by Mcat12** [comment]
There is a difference between returning a default for a given input and setting a return value for `Any` inputs. The former will only return that default value for the specified input, whereas the latter will always return a default for any input.

I was wondering about `Matcher` a bit as well. If we don't use it by default, we could add an annotation like `#[matcher]`:
```rust
#[mock_it]
trait MyTrait {
    #[matcher]
    fn test(&self, i32) -> i32;
}
```

That would change the mock for that function to `Mock<Matcher<i32>, i32>`.

**Comment by Mercanuis** [comment]
Hello, I wanted to inform that @Mcat12 that this is causing compilation failures in crate 0.3.0 and was wondering if that is just a matter of pulling in the latest package or if this is something that should be introduced. Thanks!

**Comment by Mcat12** [comment]
@Mercanuis Please open a new issue, and provide more detail. What Rust version are you using? What is the error you are seeing? How can we reproduce it? etc.
